proto/cosmos/base/reflection/v1beta1/reflection.proto

```diff
@@ -3,10 +3,161 @@ package cosmos.base.reflection.v1beta1;
+message TxDescriptor {
+  // fullname is the protobuf fullname of the raw transaction type (for instance the tx.Tx type)
+  // it is not meant to support polymorphism of transaction types, it is supposed to be used by
+  // reflection clients to understand if they can handle a specific transaction type in an application.
+  string fullname = 1;
+  // msgs lists the accepted application messages (sdk.ServiceMsg, sdk.Msg)
+  // NOTE: not to be confused with proto.Message types
+  repeated MsgDescriptor msgs = 2;
+}
+
+// AuthnDescriptor provides information on how to sign transactions without relying
+// on the online RPCs GetTxMetadata and CombineUnsignedTxAndSignatures
+message AuthnDescriptor {
+  // sign_modes defines the supported signature algorithm
+  repeated SigningModeDescriptor sign_modes = 1;
+}
+
+// SigningModeDescriptor provides information on a signing flow of the application
+// NOTE(fdymylja): here we could go as far as providing an entire flow on how
+// to sign a message given a SigningModeDescriptor, but it's better to think about
+// this another time
+message SigningModeDescriptor {
+  // name defines the unique name of the signing mode
+  string name = 1;
+  // number is the unique int32 identifier for the sign_mode enum
+  int32 number = 2;
+  // authn_info_provider_method_fullname defines the fullname of the method to call to get
+  // the metadata required to authenticate using the provided sign_modes
+  string authn_info_provider_method_fullname = 3;
+}
+
+// ChainDescriptor describes chain information of the application
+message ChainDescriptor {
+  // id is the chain id
+  string id = 1;
+}
+
+// CodecDescriptor describes the registered interfaces and provides metadata information on the types
+message CodecDescriptor {
+  // interfaces is a list of the registerted interfaces descriptors
+  repeated InterfaceDescriptor interfaces = 1;
+}
+
+// InterfaceDescriptor describes the implementation of an interface
+message InterfaceDescriptor {
+  // fullname is the name of the interface
+  string fullname = 1;
+  // interface_accepting_messages contains information regarding the proto messages which contain the interface as google.protobuf.Any field
+  repeated InterfaceAcceptingMessageDescriptor interface_accepting_messages = 2;
+  // interface_implementers is a list of the descriptors of the interface implementers
+  repeated InterfaceImplementerDescriptor interface_implementers = 3;
+}
+
+// InterfaceImplementerDescriptor describes an interface implementer
+message InterfaceImplementerDescriptor {
+  // fullname is the protobuf queryable name of the interface implementer
+  string fullname = 1;
+  // type_url defines the type URL used when marshalling the type as any
+  // this is required so we can provide type safe google.protobuf.Any marshalling and
+  // unmarshalling, making sure that we don't accept just 'any' type
+  // in our interface fields
+  string type_url = 2;
```

Comment:
I'm confused in general by your usage of both `fullname` and `type_url` here and in other places. From my perspective, there is only "type URL" OR "full name" in the protobuf world and we should use that consistently. There should not be cases where we have both a type URL and full name that could be different but refer to the same thing.

Reply:
@aaronc we do need both unfortunately.

When we build a protoregistry from gRPC reflection, the file descriptor will define, for example, the fullname of MsgSend as `cosmos.bank.v1beta1.MsgSend`, now if I want to resolve that type if it's wrapped as `google.protobuf.Any` I need to feed my protov2 type registry a way to give me my `proto.MessageType` from the given type URL which is going to be `/cosmos.bank.v1beta1.MsgSend`. 

Note: when I register a dynamic type from a file descriptor in the type registry, I cannot use a custom name. It's just going to take my MessageDescriptor from my FileDescriptor which defines its name. I don't want to edit raw file descriptors to make the (custom) any type URL match with the fullname.

Reply:
This is not quite right I believe. As I said before `/cosmos.bank.v1beta1.Msg/Send` is the type URL for the method. `cosmos.bank.v1beta1.MsgSend` is the type URL for the _request type_. So what you want to register in the Any resolver is the request type of the service method. It is _incorrect_ to say that the " fullname of MsgSend as `cosmos.bank.v1beta1.MsgSend`".

Reply:
Then I'm not understanding sorry, as per ADR-031, I need to encode my MsgBankSend as:
`any{typeURL: /cosmos.bank.v1beta1.Msg/Send, value: proto.Marshal(MsgBankSend)}`

Reply:
This is the quote from ADR-031: 
```
In order to encode service methods in transactions, we encode them as Anys in the same TxBody.messages field as other Msgs. We simply set Any.type_url to the full-qualified method name (ex. /cosmos.gov.Msg/SubmitProposal) and set Any.value to the protobuf encoding of the request message (MsgSubmitProposal in this case).
``

Reply:
Yeah, you encode and decode the request type of the method which happens to be `cosmos.v1beta1.bank.MsgSend`. That's the convention of ADR 031. We write `rpc(RequestType) returns (ResponseType)` for service methods. So you just need:
* method type URL
* request type URL
* response type URL

And you encode `Any(TypeURL: methodTypeUrl, Value: proto.Marshal(requestType))`

Does that help?